external/emulation/emulation/zhao_carr.py

```diff
@@ -226,3 +235,109 @@ def enforce_conservative_phase_dependent(state, emulator):
+    """convert data proportional to kg/kg -> kg/m2"""
+    return x * delp / GRAVITY
+
+
+def mass_to_mixing_ratio(x: np.ndarray, delp: np.ndarray) -> np.ndarray:
+    """convert data proportional to kg/m2 -> kg/kg"""
+    return x / delp * GRAVITY
+
+
+def liquid_water_equivalent(x: np.ndarray) -> np.ndarray:
+    """convert data proportional to kg/m2 -> m"""
+    return x / RHO_WATER
+
+
+def _strict_conservative_precip_from_TOA_to_surface(
+    condensate_to_precip: np.ndarray, precip_to_vapor: np.ndarray
+):
+
+    """
+    Iterates backwards through precip source and evaporation terms to
+    determine surface precip.  Functio limits evaporation to available
+    precipitation at each level, and limits source/sink terms to be positive.
+
+    Expects data in [feature x sample] dimensions (default for fields from Fortran)
+    """
+
+    limited_c_to_p = np.maximum(condensate_to_precip, 0)  # no condensate from precip
+    limited_p_to_v = np.maximum(precip_to_vapor, 0)  # no precip from vapor
+
+    if not len(precip_to_vapor.shape) == 2:
+        raise ValueError(
+            "Expected 2D inputs to the strict conservative precip function"
+        )
+
+    num_features, num_samples = precip_to_vapor.shape
+
+    total_precip = np.zeros(num_samples)
+
+    # calculate precip and evaporation starting from TOA
+    for k in range(num_features - 1, -1, -1):
+        precip = limited_c_to_p[k]
+        total_precip += precip
+        evaporation = limited_p_to_v[k]
+        limited_evap = np.minimum(total_precip, evaporation)
+        total_precip -= limited_evap
+        limited_p_to_v[k, :] = limited_evap
+
+    return limited_c_to_p, limited_p_to_v, total_precip
+
+
+def enforce_conservative_precpd(state, emulator):
+    cloud_change = emulator[PrecpdOutput.cloud_water] - state[GscondOutput.cloud_water]
+    humidity_change = emulator[PrecpdOutput.humidity] - state[GscondOutput.humidity]
+
+    # switch to kg / m2
+    delp = state[Input.delp]
+    precip_source = mixing_ratio_to_mass(-1 * cloud_change, delp)
+    precip_sink = mixing_ratio_to_mass(humidity_change, delp)
+
+    [
+        precip_src_limited,
+        precip_sink_limited,
+        total_precip,
+    ] = _strict_conservative_precip_from_TOA_to_surface(precip_source, precip_sink)
+
+    surface_precip_m = liquid_water_equivalent(total_precip)
+    limited_evaporation = mass_to_mixing_ratio(precip_sink_limited, delp)
+
+    # temperature adjust
+    evaporative_cooling = LV / CP * -1 * limited_evaporation
+
+    cloud_out = state[GscondOutput.cloud_water] + mass_to_mixing_ratio(
+        -1 * precip_src_limited, delp
+    )
+    humidity_out = state[GscondOutput.humidity] + limited_evaporation
+    temperature_out = state[GscondOutput.temperature] + evaporative_cooling
+
+    return {
+        **emulator,
+        PrecpdOutput.cloud_water: cloud_out,
+        PrecpdOutput.humidity: humidity_out,
+        PrecpdOutput.temperature: temperature_out,
```

Comment:
Is it the case then that if this flag is enabled, the original prediction of temperature change from the Precpd model is not used, and is instead overwritten by the gscond model's T prediction + cooling from the precpd-predicted cloud and q? If this might lead to unexpected behavior can this be noted in the docstrings?

Reply:
Added a note to the docstring about temperature